### DISCUSSION TITLE
How to watch route change in nuxt v3

### DISCUSSION THREAD

**Discussion by shellking4** [nuxt/nuxt#18751] [Questions]
How can I watch route change in nuxt 3
 
I have this but not woorking 

```js 
<script setup lang="ts">
watch(() => route, (first, second) => {
  console.log(
    "Watch route function called with args:",
    first,
    second
  );
});
</script>
```

**Comment by TheAlexLichter** [comment]
You can watch in e.g. `app.vue` (or layouts) as well as on the page itself if you have dynamic parameters using `watchEffect`.
See https://stackblitz.com/edit/nuxtdiscussions18751

Curious though: Where do you want to watch the route change / what are you planning to do?

**Comment by Gonzo17** [comment]
I had the same issue migrating a project from Nuxt 2 to Nuxt 3. We had something like that:
```
watch(route, loadSearchResults)
```
That worked in Nuxt 2, but not in Nuxt 3. With the `watchEffect` is does work now though. I still wonder why, I've tried to make the `watch` work with the option `{ deep: true }`, that didn't solve it. So I'm curious why the `watch` doesn't work on the `route` in Nuxt 3. Do you have an idea @manniL ?

Regarding the question of the use case, we use this on a search result page. Different components change the search by modifying the route (e.g. search type, parameters). The search page then reacts to this change and loads the data. Not sure if this best practice, there might for sure be other ways to achieve this.

**Reply by TheAlexLichter** [reply]
Hey @Gonzo17 👋🏻
Do you have a reproduction for it? 🤔

**Reply by Gonzo17** [reply]
Hey @manniL,
sorry for not replying. I'll try to provide an example later today.

**Reply by Gonzo17** [reply]
I've tried to reproduce it on a new empty project. To my surprise it actually worked with `watch` and `{ deep: true }` the same way it works with `watchEffect`. When I'm back to work and I have time, I'll try to figure out what's going wrong in our project.

**Comment by vionijannet** [comment]
I put in in my `app.vue` like this, and it works for me

```
<script setup lang="ts">
const route = useRoute();
watch(route, () => {
    // put your function in here
},  { deep: true, immediate: true });
</script>
```

If you need the old and new arguments, you can put it in the bracket after route, therefore it becomes like
`watch(route, (arg1, arg2) => { ... });`

**Comment by mkierdev** [comment]
First: You need to import route, as follows:
```js
const route = useRoute();
```

Second: Make sure not to watch whole route object. Depending on your use case, watching path is almost always enough.
```js
const route = useRoute();
watch(() => route.path, (first, second) => {
  console.log(
    "Watch route function called with args:",
    first,
    second
  );
});
```
